kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/CommitterImpl.java

```diff
@@ -0,0 +1,252 @@
+import io.tabular.iceberg.connect.events.CommitReadyPayload;
+import io.tabular.iceberg.connect.events.CommitResponsePayload;
+import io.tabular.iceberg.connect.events.Event;
+import io.tabular.iceberg.connect.events.EventType;
+import io.tabular.iceberg.connect.events.TableName;
+import io.tabular.iceberg.connect.events.TopicPartitionOffset;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.Pair;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.ListConsumerGroupOffsetsOptions;
+import org.apache.kafka.clients.admin.ListConsumerGroupOffsetsResult;
+import org.apache.kafka.clients.consumer.ConsumerGroupMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.sink.SinkTaskContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CommitterImpl implements Committer, AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CommitterImpl.class);
+  private final SinkTaskContext context;
+  private final IcebergSinkConfig config;
+  private final boolean isLeader;
+  private final CoordinatorThread coordinatorThread;
+  private final CommitRequestListener commitRequestListener;
+  private final UUID producerId;
+  private final Producer<String, byte[]> producer;
+
+  public CommitterImpl(SinkTaskContext context, IcebergSinkConfig config) {
+    this(
+        context,
+        config,
+        new ConsumerFactoryImpl(),
+        new AdminFactoryImpl(),
+        new ProducerFactoryImpl());
+  }
+
+  private CommitterImpl(
+      SinkTaskContext context,
+      IcebergSinkConfig config,
+      ConsumerFactory consumerFactory,
+      AdminFactory adminFactory,
+      ProducerFactory producerFactory) {
+    this(
+        context,
+        config,
+        new ControlTopicCommitRequestListener(config, consumerFactory),
+        adminFactory,
+        producerFactory,
+        new CoordinatorThreadFactoryImpl(adminFactory, consumerFactory, producerFactory));
+  }
+
+  @VisibleForTesting
+  CommitterImpl(
+      SinkTaskContext context,
+      IcebergSinkConfig config,
+      CommitRequestListener commitRequestListener,
+      AdminFactory adminFactory,
+      ProducerFactory producerFactory,
+      CoordinatorThreadFactory coordinatorThreadFactory) {
```

Comment:
Why does the coordinator thread need to be inside of the CommitterImpl?  Feels awkward.  All this class is doing is a bit of lifecycle work around it (other than the throw, which I'll ask in another question) at the cost of this weird null everywhere. 

Also why do we have the check for `throwExceptionIfCoordinatorThreadIsTerminated`?

Reply:
> Why does the coordinator thread need to be inside of the CommitterImpl? Feels awkward. 

Where else would you put it? This is a `CommitterImpl` detail and if `Committer` is going to be pluggable in the future, it has to be somewhere inside `CommitterImpl`?

> at the cost of this weird null everywhere.

I was trying to stay faithful to the original code but I've tried to make things a little more palatable by moving the leader election logic inside `CoordinatorThreadFactory` and make that return an `Optional<CoordinatorThread>`. Should be clearer now.

Reply:
> Also why do we have the check for `throwExceptionIfCoordinatorThreadIsTerminated`?

If you're asking why we throw an exception ...
If the CoordinatorThread is terminated, the connector will run forever and do effectively nothing. No CoordinatorThread means no commit requests being sent to the writers and no files being committed to iceberg. So we check if the CoordinatorThread is still alive and if not, fail the task, force a rebalance, and cause a new CoordinatorThread to be brought up (another option might be to just create a new CoordinatorThread but I'm trying to avoid unnecessary changes for now). 

If you're asking why would the CoordinatorThread be terminated ... I don't know but I would venture a guess this has something to do with Kafka Connect connector lifecycle madness.